lib/mongoid/persistence_context.rb

```diff
@@ -184,9 +187,19 @@ class << self
       #
       # @since 6.0.0
       def set(object, options_or_context)
-        context = PersistenceContext.new(object, options_or_context.is_a?(PersistenceContext) ?
-                                                   options_or_context.options : options_or_context)
-        Thread.current["[mongoid][#{object.object_id}]:context"] = context
+        key = "[mongoid][#{object.object_id}]:context"
+        existing_context = Thread.current[key]
+        existing_options = if existing_context
+          existing_context.options
+        else
+          {}
+        end
+        if options_or_context.is_a?(PersistenceContext)
+          options_or_context = options_or_context.options
+        end
+        new_options = existing_options.merge(options_or_context)
+        context = PersistenceContext.new(object, new_options)
+        Thread.current[key] = context
```

Comment:
Out of curiosity, will this cause issues if there's more code after the inner block within the outer block? e.g.

```ruby
Band.with(collection: "artists") do |klass|
  Band.with(client: :reports) do |klass|
    klass.new.collection_name
  end

  klass.new.collection_name
end
```ruby

Reply:
Yes, repaired this as well.